primary/src/proposer.rs

```diff
@@ -201,27 +224,63 @@ impl Proposer {
+                self.last_parents.extend(parents)
+            }
+        }
+
+        // Check whether we are on the happy path (we can vote for the leader or the leader has enough
+        // votes to enable a commit).
+        self.common_case = self.is_common_case();
+    }
+
     /// Main loop listening to incoming messages.
     pub async fn run(&mut self) {
         debug!("Dag starting at round {}", self.round);
-        let mut advance = true;
 
-        let timer = sleep(self.max_header_delay);
+        let timer = sleep(self.delta);
         tokio::pin!(timer);
 
         info!("Proposer on node {} has started successfully.", self.name);
         loop {
-            // Check if we can propose a new header. We propose a new header when we have a quorum of parents
-            // and one of the following conditions is met:
-            // (i) the timer expired (we timed out on the leader or gave up gather votes for the leader),
-            // (ii) we have enough digests (minimum header size) and we are on the happy path (we can vote for
-            // the leader or the leader has enough votes to enable a commit). The latter condition only matters
-            // in partially synchrony.
-            let enough_parents = !self.last_parents.is_empty();
-            let enough_digests = self.payload_size >= self.header_size;
+            // Check if we have enough parents; if we have enough digests; if other primaries already
+            // made a non-certificate for this round (and need our certificate to make progress); and
+            // if the timer expired (only meaningful in partial-synchrony).
             let timer_expired = timer.is_elapsed();
-
-            if (timer_expired || (enough_digests && advance)) && enough_parents {
+            let enough_parents = !self.last_parents.is_empty();
+            let we_have_enough_batches = self.payload_size >= self.header_size;
+            let others_have_batches = self
+                .highest_peers_round
+                .is_some_and(|r| self.round == r - 1);
+
+            // Check whether advancing round allows the system to make meaningful progress. The system
+            // makes meaningful progress if it advances with non-empty certificates.
+            let meaningful_progress = we_have_enough_batches || others_have_batches;
+
+            // Print a few debug logs.
+            debug!("Do we have enough parents: {enough_parents}");
+            debug!("Do we have enough digests: {we_have_enough_batches}");
+            debug!("Have we learned about a non-empty certificate for this round: {others_have_batches}");
+            debug!("Are we in the common case: {}", self.common_case);
+            debug!("Did the timer expire: {timer_expired}");
+
+            // Check whether we can move to the next round and proposer a new header.
+            if (self.common_case || timer_expired) && enough_parents && meaningful_progress {
```

Comment:
So we go from 1.
```rust
(timer_expired || (we_have_enough_digests && common_case)) && enough_parents
```
to 2.
```rust
(common_case || timer_expired) && enough_parents && (we_have_enough_digests || others_have_batches)
```

1. is equivalent to
```rust
(common_case || timer_expired) && enough_parents && (we_have_enough_digests || timer_expired)
```
Which means you're just replacing `timer_expired` by `others_have_batches` in the third clause.

Fair enough.

Now here's my problem: assume everybody else, for some reason, proposes a valid but empty header, and I as a validator do not have batches. Nobody proposes. Then the others start receiving transactions.

Don't we lose liveness?

Can we accuse a number of other nodes of something specific? Which nodes?

I think we might need another timer.

Reply:
You are right, if this buggy behaviour happens I don't see how we can create a fraud proof. However I don't think we have fraud proofs for liveness attacks anyway (e.g. what do we do if buggy nodes simply stop creating headers?).

Let's keep thinking about this (I would not recommend landing this PR until we are sure DevNet works anyway).

Proposing as soon as f+1 nodes propose empty or non-empty header may not work. Upon booting, there are 2f+1 empty certificates (genesis); this means all honest nodes will propose their first (possibly empty) header; from there it seems they will always have 2f+1 (possibly empty) headers triggering nodes to move forward.

Reply:
> Proposing as soon as f+1 nodes propose empty or non-empty header may not work. Upon booting, there are 2f+1 empty certificates (genesis); this means all honest nodes will propose their first (possibly empty) header; from there it seems they will always have 2f+1 (possibly empty) headers triggering nodes to move forward.

You seem to be reasoning on empty rounds in the round that precedes the proposal (the parents). I'm suggesting reasoning on counting the proposals of the current round.

Let me put it this way, you're proposing this for gating the proposal condition:
```
(common_case || timer_expired) && enough_parents && (we_have_enough_digests || others_have_batches)
```

Could we not recover from the case where all "committed to empty" with the following condition:
```
(common_case || timer_expired) && enough_parents && (we_have_enough_digests || others_have_batches || (timer_expired && f_plus_1_proposed_this_round))
```
?

Reply:
Ha I see. Yes, I believe this would fix the scenario you described

Reply:
I believe this 6-variable clause will need really, really didactic commenting around it :D

Reply:
So this will fix the issue you described, but it also mean that if there are no transactions we keep making empty certificates? (Although at a lower pace, dictated by 'timer_expired').

Let's say there are no client transactions in the system and that all authorities are honest.
At round 0:
-  Once the first timer expires: (common_case || timer_1_expired) = true
- Since all nodes have the genesis: enough_parents = true
- when the second timer expires, since all nodes have f+1 genesis blocks (round 0): (timer_2_expired && f_plus_1_proposed_this_round)) = true

Reply:
> since all nodes have f+1 genesis blocks
How is that leading to `f_plus_1_proposed_this_round == true`?

Reply:
The current round is round 0; do you mean the "next_round"?

Reply:
Yes.